commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/stash/StashPublisher.java

```diff
@@ -121,7 +127,16 @@ private void vote(@NotNull SBuild build,
                     @NotNull StashBuildStatus status,
                     @NotNull String comment) {
     String msg = createMessage(status, build.getBuildPromotion().getBuildTypeExternalId(), getBuildName(build), myLinks.getViewResultsUrl(build), comment);
-    vote(revision.getRevision(), msg, LogUtil.describe(build));
+    String commitRevision = revision.getRevision();
+    try {
+      StashCommit commit = findPullRequestCommit(build);
```

Comment:
I am not sure if this can be addressed easily, but at the moment we try to issue HTTP requests outside the threads where the event handlers code works, so the event handlers return as quick as possible. See postAsync(...) calls in vote(...). The findPullRequestCommit(...) call is performed in the same thread and will make buildStarted, buildFinished etc. waiting if the branch name contains "merge".